### ISSUE TITLE
3.5.0 fails while deploying to multiple stages in parallel

### ISSUE THREAD

**Issue by ledermann** [capistrano/capistrano#1649]
#### Steps to reproduce

1. Add `gem 'capistrano', '3.5.0'` to your Gemfile
2. Execute `bundle exec cap install STAGES=production,staging`
3. Execute `bundle exec cap production deploy` (works fine)
4. Execute `bundle exec cap staging deploy` (works fine)
5. Execute `bundle exec cap production staging deploy` (fails)

#### Expected behaviour

Deployment should work on multiple stages in parallel.

#### Actual behaviour

With Capistrano v3.5.0 deploy command fails with error message `Using can't modify frozen object` in [lib/capistrano/setup.rb:25](https://github.com/capistrano/capistrano/blob/v3.5.0/lib/capistrano/setup.rb#L25). 

It seems to be a regression of #1452. In v3.4.1 it works fine.

#### My configuration

```
$ bundle exec cap production staging doctor --trace
** Invoke production (first_time)
** Execute production
** Invoke load:defaults (first_time)
** Execute load:defaults
** Invoke bundler:map_bins (first_time)
** Invoke passenger:bundler:hook (first_time)
** Execute passenger:bundler:hook
** Execute bundler:map_bins
** Invoke deploy:set_rails_env (first_time)
** Execute deploy:set_rails_env
** Invoke deploy:set_linked_dirs (first_time)
** Execute deploy:set_linked_dirs
** Invoke deploy:set_rails_env
** Invoke staging (first_time)
** Execute staging
** Invoke load:defaults
cap aborted!
can't modify frozen object
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/capistrano-3.5.0/lib/capistrano/setup.rb:25:in `extend_object'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/capistrano-3.5.0/lib/capistrano/setup.rb:25:in `extend'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/capistrano-3.5.0/lib/capistrano/setup.rb:25:in `block (2 levels) in <top (required)>'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/task.rb:248:in `call'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/task.rb:248:in `block in execute'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/task.rb:243:in `each'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/task.rb:243:in `execute'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/airbrussh-1.0.1/lib/airbrussh/rake/context.rb:55:in `execute'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/task.rb:187:in `block in invoke_with_call_chain'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/task.rb:180:in `invoke_with_call_chain'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/task.rb:173:in `invoke'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/application.rb:150:in `invoke_task'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/application.rb:106:in `block (2 levels) in top_level'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/application.rb:106:in `each'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/application.rb:106:in `block in top_level'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/application.rb:115:in `run_with_threads'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/application.rb:100:in `top_level'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/application.rb:78:in `block in run'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/application.rb:176:in `standard_exception_handling'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/rake-11.1.2/lib/rake/application.rb:75:in `run'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/capistrano-3.5.0/lib/capistrano/application.rb:14:in `run'
/Users/ledermann/.rbenv/versions/2.1.7/lib/ruby/gems/2.1.0/gems/capistrano-3.5.0/bin/cap:3:in `<top (required)>'
/Users/ledermann/.rbenv/versions/2.1.7/bin/cap:22:in `load'
/Users/ledermann/.rbenv/versions/2.1.7/bin/cap:22:in `<main>'
Tasks: TOP => staging
```

**Comment by leehambley** [comment]
Thanks for the report @ledermann I'm wondering where you found documented that `STAGES=foo,bar` was supposed to a parallel deployment, I might have missed the memo but I *thought* stages were only supposed to be used this way when running `cap install` the first time.

It's absolutely OK if this *becomes* a use-case (or, if I missed a memo) but I'm not sure we ever had documented this as a valid way for things to work.

**Comment by ledermann** [comment]
Hm, it seems that in the Rails app I'm just recently working on Capistrano was used in an undocumented way: Instead of defining multiple servers in `config/deploy/production.rb` there were separate config files in `config/deploy` for every server (just containing the `server` line and setting the `:rails_env` to `production`). Deployment was executed by doing `bundle exec cap production staging deploy` to deploy to the first server `production` and the second server `staging` in parallel. The concept of "stage" was mixed up with the concept of "server". Nevertheless, this works fine until 3.4.1, but fails in 3.5.0.

Now, after combining the multiple stage files into a single one (having multiple `server` lines) it works fine with 3.5.0.

So, no bug at all. Thanks for advice, @leehambley :)

**Comment by leehambley** [comment]
> Now, after combining the multiple stage files into a single one (having multiple server lines) it works fine with 3.5.0.

I'm glad! I'd be more than happy to have the issue re-opened if you do think this is a legit and valuable use-case, but I think it was just a mis-setup.

I don't want you to go away with a worse workflow/experience because of a mis-feature being accidentally rendered ineffective.

**Comment by ledermann** [comment]
I my case it was just a mis-setup. But because deploying to multiple stages in parallel was an undocumented feature in 3.4.1 which is not possible anymore, maybe it's your decision to enable it as an official feature.

**Comment by leehambley** [comment]
I think I'd prefer to let people run that in parallel using their shell, I can't imagine what multi-threading something already multi-threaded might do!

With GNU Parallel (widely available on linux, and via homebrew) you could achieve that with:

    $ parallel -j2 bundle exec cap {} deploy ::: staging production

(which runs `bundle exec cap __ deploy` twice (`-j2`) using the inputs `staging` and `production` as placeholders for the `{}`)

**Comment by ledermann** [comment]
Yes, using GNU Parallel could be a replacement, although the output is not as beautiful (means: merged) as with plain Capistrano (I used v3.4.1 with Airbrussh).

Nevertheless: For me it's ok to fix my setup. I'm curious if there are other people who used the undocumented feature and will complain about that it's missing in v3.5.0.

Thank you very much for helping out!

**Comment by leehambley** [comment]
> Yes, using GNU Parallel could be a replacement, although the output is not as beautiful (means: merged) as with plain Capistrano (I used v3.4.1 with Airbrussh).

Absolutely! I often wondered why there wasn't some way to start a tmux session with `n` splits, and have the output streamed to each one, I guess with some pipe trickery it could be done but that strays away from the path of "hah, cool hack bro 🤘🏻 " into "omfg wtf have u done 👹"

**Comment by painkillereldad** [comment]
Hi,

getting the same for a usual deploy to staging using the staging.rb .

i do cap staging deploy  and i get 

** Invoke production (first_time)
** Execute production
** Invoke load:defaults (first_time)
** Execute load:defaults
** Invoke bundler:map_bins (first_time)
** Execute bundler:map_bins
** Invoke deploy:set_rails_env (first_time)
** Execute deploy:set_rails_env
** Invoke deploy:set_linked_dirs (first_time)
** Execute deploy:set_linked_dirs
** Invoke deploy:set_rails_env
** Invoke staging (first_time)
** Execute staging
** Invoke load:defaults
cap aborted!
can't modify frozen object

never encountered this before any idea

**Comment by mattbrictson** [comment]
@painkillereldad Your trace shows that both `production` *and* `staging` are being invoked. 

```
** Invoke production (first_time)
...
** Invoke staging (first_time)
```

That is a problem, because Capistrano can only activate one stage at a time. If you run `cap staging deploy`, then only staging should be activated; likewise `cap production deploy` should only activate production. (Note that `cap production staging deploy` is invalid usage.)

Have you added code to your `deploy.rb` (or elsewhere) that would cause both stages to be invoked?

If you believe there is a bug within Capistrano itself that is causing `production` to be invoked even when you run `cap staging deploy`, then please open a new issue. Include the version of Capistrano you are using, plus a copy of your `Capfile`.

Thanks!
